mbq/client/contrib/permissions.py

```diff
@@ -0,0 +1,249 @@
+    """Raised from within OSCoreClient implementations to denote the fetch
+    failed due to a server-side error.
+    """
+
+    pass
+
+
+class OSCoreClient(Protocol):
+    def fetch_permissions(
+        self, person_id: str, org_ref: UUIDType
+    ) -> FetchedPermissionsDoc:
+        ...
+
+    def fetch_permissions_for_location(
+        self, person_id: str, location_id: int, location_type: RefType
+    ) -> FetchedPermissionsDoc:
+        ...
+
+
+class OSCoreServiceClient:
+    def __init__(self, client: ServiceClient):
+        # The copying and munging here is attempting to deal with differences
+        # in how the individual ServiceClients are configured. We can get rid
+        # of it if we standardize within the services.
+
+        # Note that we are doing a shallow copy so the Authenticator instance
+        # will be shared
+        self.client = copy(client)
+
+        def _post_process_response(data):
+            if isinstance(data, dict):
+                objects = data.get("objects", data)
+                objects["global"] = []
+                return objects
+            else:
+                return data
+
+        self.client._post_process_response = _post_process_response
+        self.client._headers = None
+
+        parsed = urllib.parse.urlparse(self.client._api_url)
+        self.client._api_url = f"{parsed.scheme}://{parsed.netloc}"
+
+    def fetch_permissions(
+        self, person_id: str, org_ref: UUIDType
+    ) -> FetchedPermissionsDoc:
+        logger.debug(f"Fetching permissions from OS Core: {person_id}, {org_ref}")
+
+        try:
+            return self.client.get(
+                f"/api/v1/people/{person_id}/permissions", params={"org_ref": org_ref}
+            )
+        except requests.exceptions.HTTPError as e:
+            response = getattr(e, "response", None)
+            if response and response.status_code // 100 == 4:
+                raise ClientError("Invalid request") from e
+            raise ServerError("Server error") from e
+        except Exception as e:
+            raise ServerError("Server error") from e
+
+    def fetch_permissions_for_location(
+        self, person_id: str, location_id: int, location_type: RefType
+    ) -> FetchedPermissionsDoc:
+        logger.debug(
+            f"Fetching permissions from OS Core: {person_id}, {location_type} {location_id}"
+        )
+
+        try:
+            return self.client.get(
+                f"/api/v1/people/{person_id}/permissions",
+                params={"location_id": location_id, "location_type": location_type},
+            )
+        except requests.exceptions.HTTPError as e:
+            response = getattr(e, "response", None)
+            if response and response.status_code // 100 == 4:
+                raise ClientError("Invalid request") from e
+            raise ServerError("Server error") from e
+        except Exception as e:
+            raise ServerError("Server error") from e
```

Comment:
The exception handling is copy-pasta'd, don't think it's worth the complexity to share it